consensus/propagation/reactor.go

```diff
@@ -135,3 +158,255 @@ func (blockProp *Reactor) Receive(chID byte, peer p2p.Peer, msgBytes []byte) {
+	}
+	return peers
+}
+
+// setPeer sets the peer state for the given peer.
+//
+//nolint:unused
+func (blockProp *Reactor) setPeer(peer p2p.ID, state *PeerState) {
+	blockProp.mtx.Lock()
+	defer blockProp.mtx.Unlock()
+	blockProp.peerstate[peer] = state
+}
+
+// ProposeBlock is called when the consensus routine has created a new proposal
+// and it needs to be gossiped to the rest of the network.
+func (blockProp *Reactor) ProposeBlock(proposal *types.Proposal, haves *bits.BitArray) {
+	blockProp.HandleProposal(proposal, blockProp.self, haves)
+}
+
+// HandleProposal adds a proposal to the data routine. This should be called any
+// time a proposal is received from a peer or when a proposal is created. If the
+// proposal is new, it will be stored and broadcast to the relevant peers.
+// This function will:
+// - check if the from peer is provided. If not, set it to self.
+// - add the proposal to the reactor's state.
+// - if adding the proposal to the state is successful, broadcast the proposal to the peers.
+// Note: this method will not propagate the haves after the proposal and the compact block is propagated.
+// Check broadcastSelfProposalHaves for that.
+func (blockProp *Reactor) HandleProposal(proposal *types.Proposal, from p2p.ID, haves *bits.BitArray) {
+}
+
+// broadcastProposal gossips the provided proposal to all peers. This should
+// only be called upon receiving a proposal for the first time or after creating
+// a proposal block.
+//
+//nolint:unused
+func (blockProp *Reactor) broadcastProposal(proposal *types.Proposal, from p2p.ID) {
+}
+
+// broadcastSelfProposalHaves broadcasts the haves to all the connected peers when we're the proposers.
+// Note: the haves are chunked so that every peer only receives a portion of the haves.
+//
+//nolint:unused
+func (blockProp *Reactor) broadcastSelfProposalHaves(proposal *types.Proposal, from p2p.ID, haves *bits.BitArray) {
+}
+
+// chunkParts takes a bit array then returns an array of chunked bit arrays.
+// TODO document how the redundancy and the peer count are used here.
+func chunkParts(p *bits.BitArray, peerCount, redundancy int) []*bits.BitArray {
+	size := p.Size()
+	if peerCount == 0 {
+		peerCount = 1
+	}
+	chunkSize := size / peerCount
+	// round up to use the ceil
+	if size%peerCount != 0 || chunkSize == 0 {
+		chunkSize++
+	}
+
+	// Create empty bit arrays for each peer
+	parts := make([]*bits.BitArray, peerCount)
+	for i := 0; i < peerCount; i++ {
+		parts[i] = bits.NewBitArray(size)
+	}
+
+	chunks := chunkIndexes(size, chunkSize)
+	cursor := 0
+	for p := 0; p < peerCount; p++ {
+		for r := 0; r < redundancy; r++ {
+			start, end := chunks[cursor][0], chunks[cursor][1]
+			for i := start; i < end; i++ {
+				parts[p].SetIndex(i, true)
+			}
+			cursor++
+			if cursor >= len(chunks) {
+				cursor = 0
+			}
+		}
+	}
+
+	return parts
+}
+
+// chunkIndexes
+// TODO document and explain the parameters
+func chunkIndexes(totalSize, chunkSize int) [][2]int {
```

Comment:
this is just used in the above function to split up the block into equal portions